### PR TITLE
Add `preserveConsecutiveUppercase` option

### PR DESCRIPTION
**Pull request by donkeyDau**
Closes https://github.com/sindresorhus/decamelize/issues/29

Names of options are open for discussion (as everything else as well :) ).

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -1,3 +1,19 @@
...
+		Preserve blocks of uppercase characters: `fooBAR` → `foo_BAR`.
+		@default false
+		*/
+		readonly preserveConsecutiveUppercase?: boolean;
+
+		/**
+		Last character of uppercase block starts new section: `fooBARest` → `foo_BA_rest`.
+		@default false
+		*/
+		readonly consecutiveUppercaseNotGreedy?: boolean;
```
Comment (sindresorhus) [inline index.d.ts:13]: This was not mentioned in the issue. Why is it even an option? I think this should just be `true` by default.

Comment (donkeyDau) [reply index.d.ts:13]: This came to my mind during implementation: How to deal with the last character of an uppercase block? As you seem not to be happy with it: For my usecase there's no need for this option. We can drop it and set the behavior to `true`.

Please tell me how you want to deal with this option.

Comment (sindresorhus) [reply index.d.ts:13]: Yes, drop it and make it `true` by default.

File: index.js [hunk 2]
```diff
@@ -1,12 +1,85 @@
...
+		prefix = '(?<!^)';
+	}
+
+	const splitUpperLowerRegExp = new RegExp(
+		prefix +
+			'(\\p{Uppercase_Letter}+)(\\p{Uppercase_Letter}\\p{Lowercase_Letter}+)',
+		'gu'
+	);
+
+	// If second char is uppercase -> we have a block at start
```
Comment (sindresorhus) [inline index.js:22]: "chars" => "characters"

Applies everywhere.

File: index.js [hunk 3]
```diff
@@ -1,12 +1,85 @@
...
+		});
+	}
+
+	return decamelized.replace(
+		/([\p{Uppercase_Letter}\d]+)(\p{Lowercase_Letter})/gu,
+		`$1${separator}$2`
+	);
+};
+
+module.exports = (text, separator = '_', options) => {
```
Comment (sindresorhus) [inline index.js:48]:
```suggestion
module.exports = (text, separator = '_', {preserveConsecutiveUppercase = false} = {}) => {
```

File: index.js [hunk 4]
```diff
@@ -1,12 +1,85 @@
...
 	if (!(typeof text === 'string' && typeof separator === 'string')) {
-		throw new TypeError('The `text` and `separator` arguments should be of type `string`');
+		throw new TypeError(
+			'The `text` and `separator` arguments should be of type `string`'
+		);
```
Comment (sindresorhus) [inline index.js:52]: Don't do unrelated changes.

Comment (donkeyDau) [reply index.js]: Sure. My IDE was a bit overhasty ...

**Comment by donkeyDau** [comment]
I've implemented the way you proposed. Please re-check. Thx!

File: index.d.ts [hunk 5]
```diff
@@ -1,3 +1,14 @@
+declare namespace decamelize {
+	interface Options {
+		/**
+		Preserve blocks of uppercase characters: `fooBARAction` → `foo_BAR_action`.
+
```
Comment (sindresorhus) [inline index.d.ts:5]: Can you add an example code block here instead and show an example with and without the option as `true`?

File: index.js [hunk 6]
```diff
@@ -1,12 +1,52 @@
...
+const handlePreserveConsecutiveUppercase = (decamelized, separator) => {
+	// Lowercase all non-block uppercase characters.
+	decamelized = decamelized.replace(
+		/((?<![\p{Uppercase_Letter}\d])[\p{Uppercase_Letter}\d](?![\p{Uppercase_Letter}\d]))/gu,
+		$0 => {
+			return $0.toLowerCase();
+		}
+	);
+
+	// Use last uppercase character as starting of a new section followed lowercase character(s).
```
Comment (sindresorhus) [inline index.js:10]: I think this comment could be written clearer.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Also, review your code and see if it can be simplified somehow. I often comment this as many times there are ways to simplify the code when looking it over a couple of times.

**Comment by sindresorhus** [comment]
Can you also add a test to ensure none of the regexes have catastrophic backtracking? (Meaning running the function on a really large generated input string). I'm a bit paranoid these days about that, as I've had some issues with it in the past.

**Comment by donkeyDau** [comment]
> Can you also add a test to ensure none of the regexes have catastrophic backtracking? (Meaning running the function on a really large generated input string). I'm a bit paranoid these days about that, as I've had some issues with it in the past.

I set up some tests. Didn't see any huge performance impact. Don't know if the string length is sufficient ...

> Also, review your code and see if it can be simplified somehow. I often comment this as many times there are ways to simplify the code when looking it over a couple of times.

I meditated over the code for some time but didn't see any improvement possibility. One could "walk" the string character by character and determine if a separator should be inserted or not but I don't think this is the way to go.

File: test.js [hunk 7]
```diff
@@ -28,6 +31,93 @@ test('handles acronyms', t => {
...
+		decamelize('CAPLOCKED1', {
+			preserveConsecutiveUppercase: true
+		}),
+		'CAPLOCKED1'
+	);
+});
+
+test('long strings', t => {
+	const longString =
+		'Lb8SvAARMshcNvfxjgGCgfot3AZAzysuxRpG9XfpLCz89TeWqAd3TUo64K45VH2MfjLYhztt4LQYzrEbTpx7gGcG4T8ueKPm6VraXKtULJdncFQhEQfCRwWGNscdFe6UTEAvN7Nze4Qy4hvZuKLX5YiohGpvNZUtLGen3WP2jot8VeprzyXQmiKdxdxrEResSRgSWENCzXZPSerYuEfApVbjuDJZ9kGMRXFRZQVyBDDGfY9ERqtxHQxPw65TtEo3dgwhcuhvC3dMyRJ6jWaonKB3Pqtv27vRv5MgYb5mgvCE55oCTBG9yASPaw2KqYVz3amBge9HggEzXJGhwSXjkL7jUYk3WjQUbwVnZNHkH3P9MpvM98DtTnGAYfK5TjD8Y5oXPRJmdCHzhByboaW2oRJ2Ft7dxGKXLs2s7qsQs8FsJHVcYrmVHRa6th5CizHSXK7vr5D3KYsfsnr92AmtR4LERam7CV9emBBuykQJMejLGFsvgTrBKmmUqijxSgY';
```
Comment (sindresorhus) [inline test.js:106]: Can you do `.repeat(100)` on this?

File: index.d.ts [hunk 8]
```diff
@@ -1,8 +1,47 @@
+declare namespace decamelize {
+	interface Options {
+		/**
+		The separator to use to put in between the words from `string`.
```
Comment (sindresorhus) [inline index.d.ts:4]: I think this sentence could be better written. It should also be in the readme.

File: test.js (hunk 7 above)
Comment (donkeyDau) [reply test.js]: Done.

File: index.d.ts (hunk 8 above)
Comment (donkeyDau) [reply index.d.ts]: I gave it try. The current sentence was just a copy&paste from the existing docs.

**Comment by sindresorhus** [comment]
Thanks for your good work on this 👍🏻

**Comment by donkeyDau** [comment]
The thanks is all mine. Thank you for creating and maintaining!
